core/src/main/java/org/infinispan/interceptors/impl/TransactionalStoreInterceptor.java

```diff
@@ -0,0 +1,158 @@
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.interceptors.DDAsyncInterceptor;
+import org.infinispan.marshall.core.MarshalledEntryImpl;
+import org.infinispan.persistence.PersistenceUtil;
+import org.infinispan.persistence.manager.PersistenceManager;
+import org.infinispan.persistence.support.BatchModification;
+
+import javax.transaction.Transaction;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import static org.infinispan.factories.KnownComponentNames.CACHE_MARSHALLER;
+import static org.infinispan.persistence.manager.PersistenceManager.AccessMode.SHARED;
+
+/**
+ * An interceptor which ensures that writes to an underlying transactional store are prepared->committed/rolledback as part
+ * of the 2PC, therefore ensuring that the cache and transactional store(s) remain consistent.
+ *
+ * @author Ryan Emerson
+ * @since 9.0
+ */
+public class TransactionalStoreInterceptor extends DDAsyncInterceptor {
+   private PersistenceManager persistenceManager;
+   private InternalEntryFactory entryFactory;
+   private StreamingMarshaller marshaller;
+   private volatile boolean enabled = true;
+
+   @Inject
+   protected void init(PersistenceManager persistenceManager, InternalEntryFactory entryFactory,
+                       @ComponentName(CACHE_MARSHALLER) StreamingMarshaller marshaller) {
+      this.persistenceManager = persistenceManager;
+      this.entryFactory = entryFactory;
+      this.marshaller = marshaller;
+   }
+
+   @Override
+   public CompletableFuture<Void> visitPrepareCommand(TxInvocationContext ctx, PrepareCommand command) throws Throwable {
+      if (isValidContext(ctx)) {
+         Transaction tx = ctx.getTransaction();
+         Updater modBuilder = new Updater();
+         List<WriteCommand> modifications = ctx.getCacheTransaction().getAllModifications();
+         for (WriteCommand writeCommand : modifications) {
+            writeCommand.acceptVisitor(ctx, modBuilder);
+         }
+         persistenceManager.prepareAllTxStores(tx, modBuilder.modifications, SHARED);
+      }
+      return ctx.continueInvocation();
+   }
+
+   @Override
+   public CompletableFuture<Void> visitCommitCommand(TxInvocationContext ctx, CommitCommand command) throws Throwable {
+      if (isValidContext(ctx)) {
+         persistenceManager.commitAllTxStores(ctx.getTransaction(), SHARED);
+      }
+      return ctx.continueInvocation();
+   }
+
+   @Override
+   public CompletableFuture<Void> visitRollbackCommand(TxInvocationContext ctx, RollbackCommand command) throws Throwable {
+      if (isValidContext(ctx)) {
+         persistenceManager.rollbackAllTxStores(ctx.getTransaction(), SHARED);
+      }
+      return ctx.continueInvocation();
+   }
+
+   public void disableInterceptor() {
+      enabled = false;
+   }
+
+   private boolean isValidContext(TxInvocationContext ctx) {
+      return isStoreEnabled() && ctx.isOriginLocal() && ctx.isInTxScope();
```

Comment:
The method name is a bit misleading, the context itself can't be invalid. For the tx commands, `ctx.isInTxScope() == true`, so I'd remove that check and inline the rest of the method.